### ISSUE TITLE
How is token caching supposed to work with horizontal scaling?

### ISSUE THREAD

**Issue by justinmchase** [AzureAD/microsoft-authentication-library-for-js#4376]
### Core Library

MSAL Node (@azure/msal-node)

### Core Library Version

1.3.0

### Wrapper Library

Not Applicable

### Wrapper Library Version

None

### Description

I am using the ConfidentialClientApplication form my backend and I'm trying to support refresh tokens and its my understanding that I should instead use the acquireTokenSilent function and that msal-node does the caching... but I can't quite comprehend how this can possibly work.

It seems like the default caching is simply in-memory caching, which makes no sense for a horizontally scaled application. I went ahead and looked at the ICachePlugin interface and its quite perplexing... it seems like its expecting me to serialize _the entire_ cache and store it in a single blob. This makes no sense.

Suppose I have 100 servers scaled horizontally and a million users. They're all logging in and randomly accessing these servers and requesting refreshed tokens, how am I supposed to use this to support that scenario? I'm expecting to be able to write the cache to a database, but I don't see how... How can I expect each one of these to be writing files this frequently with all of this information? And if multiple servers are reading and writing from the cache at the same time then how can this work? It seems like it needs much more fine grained functions on the plugin, where I can essentially store every individual token in a table in a database and every request or update the a token is fetched or stored as a row in that table.

And to be honest, I don't even understand why you even trying to cache these at all? How is this better than just giving each user their own refresh token to individually keep track of? That would be a 0 cache solution and all this complexity would just go away. What is even the benefit? What am I missing here?

I'm super skeptical of this design decision and am seriously wondering if this can even work, please help me understand here, the docs and samples are not quite covering this pretty common use case.

### Identity Provider

Azure AD / MSA

### Source

External (Customer)

**Comment by samuelkubai** [comment]
Hi @justinmchase,

To answer the question of "Why cache this at all?", we cache this information to reduce the round trips to the server and back and have the subsequent sign ins be a lot faster by utilizing the cached data.

You are absolutely right about horizontal scaling that's what the pull request #4279 is working on adding, a layer to allow our users to use some form of cache like redis and the like to store this information on and correctly retrieve and store as required by multiple servers horizontally scaled.

**Comment by samuelkubai** [comment]
Once this pull request is merged it should provide the right tools and samples to make this possible.

**Comment by justinmchase** [comment]
Hello @samuelkubai  I left some code comments on the PR, sorry I could be wrong but it looks like it probably won't be sufficient. 

I managed to find a workaround which is to tap into the kvstore in the cache after token is created to find the refresh token. But it feels like a hack and I'm not real happy about getting stuff right from the kvstore like this.

On the other hand I think adding a IStoragePlugin would work, how open would you be to a PR along those lines?

**Comment by samuelkubai** [comment]
Hi @justinmchase, I have addressed your concerns on the pull request.

**Comment by justinmchase** [comment]
Would it be possible for this library to be updated to get the refreshToken directly as well? Like suppose i didn't want to use the caching in this library at all?

I don't want to have to reach into the kvstore and try to guess the cache keys... I just want to give my users their refresh token and not do any of this server side caching at all.

**Comment by justinmchase** [comment]
I've gotten the new version of the library and will attempt to implement the pattern based on the redis example. I'll add further details about concerns / successes here.

**Comment by samuelkubai** [comment]
Thank you @justinmchase, will be closing this issue for now.
Feel free to re-open it or create a new issue if you have any concerns or successes on this.
